### PR TITLE
feat(manager/pep621): support pdm lock files

### PR DESCRIPTION
**Pull request by secustor**
## Changes
This PR is a follow up to: 
- https://github.com/renovatebot/renovate/pull/22082
- https://github.com/containerbase/base/pull/973

and add support for PDM lockfiles ( `pdm.lock` ) 

## Context
https://github.com/renovatebot/renovate/issues/10187

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

https://github.com/secustor/pdm-pep621-reproduction

### CODE REVIEW

File: lib/modules/manager/pep621/processors/pdm.ts [hunk 1]
```diff
@@ -39,4 +48,75 @@ export class PdmProcessor implements PyProjectProcessor {
...
+        toolName: 'pdm',
+        constraint: config.constraints?.pdm,
+      };
+
+      const execOptions: ExecOptions = {
+        docker: {},
+        toolConstraints: [toolConstraint],
+      };
+
+      // on lockFileMaintenance do not specify any packages and update the complete lock file
```
Comment (rarkins) [inline lib/modules/manager/pep621/processors/pdm.ts:78]: Does `pdm update` update the complete lock file? In some languages we need to delete the lock file first to get it to do a proper full refresh

Comment (secustor) [reply lib/modules/manager/pep621/processors/pdm.ts:78]: Based on the documentation yes and during testing I have found nothing which contradicts that. 
https://pdm.fming.dev/2.6/reference/cli/#update_2

**Review by rarkins — DISMISSED** [review]

File: lib/modules/manager/pep621/artifacts.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import { writeLocalFile } from '../../../util/fs';
+import type { UpdateArtifact, UpdateArtifactsResult } from '../types';
+import { processors } from './processors';
+
+export async function updateArtifacts(
+  updateArtifact: UpdateArtifact
+): Promise<UpdateArtifactsResult[] | null> {
+  const { packageFileName, newPackageFileContent } = updateArtifact;
+
+  await writeLocalFile(packageFileName, newPackageFileContent);
```
Comment (viceice) [inline lib/modules/manager/pep621/artifacts.ts:11]:
```suggestion
```
@rarkins doesn't the upper layer already to this? i havn't seen such a line in other managers.

File: lib/modules/manager/pep621/processors/pdm.spec.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+      expect(result).toBeNull();
+      expect(execSnapshots).toMatchObject([
+        {
+          cmd: 'docker pull containerbase/sidecar',
+        },
+        {
+          cmd: 'docker ps --filter name=renovate_sidecar -aq',
+        },
+        {
+          cmd: 'docker run --rm --name=renovate_sidecar --label=renovate_child -v "/tmp/github/some/repo":"/tmp/github/some/repo" -v "/tmp/cache":"/tmp/cache" -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w "/tmp/github/some/repo" containerbase/sidecar bash -l -c "install-tool pdm v2.5.0 && pdm update dep1"',
```
Comment (viceice) [inline lib/modules/manager/pep621/processors/pdm.spec.ts:65]: see other comment about format

File: lib/modules/manager/pep621/processors/pdm.spec.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        },
+      ]);
+    });
+
+    it('rethrow error', async () => {
```
Comment (viceice) [inline lib/modules/manager/pep621/processors/pdm.spec.ts:70]:
```suggestion
    it('returns artifact error', async () => {
```

File: lib/modules/manager/pep621/processors/pdm.ts [hunk 5]
```diff
@@ -39,4 +48,75 @@ export class PdmProcessor implements PyProjectProcessor {
...
+      // else only update specific packages
+      let packageList = '';
+      if (!isLockFileMaintenance) {
+        packageList = updatedDeps
+          .map((value) => value.packageName ?? value.depName)
```
Comment (viceice) [inline lib/modules/manager/pep621/processors/pdm.ts:83]:
```suggestion
          .map((value) => value.packageName)
```

`packageName` is mandatory

File: lib/modules/manager/pep621/processors/pdm.ts [hunk 6]
```diff
@@ -39,4 +48,75 @@ export class PdmProcessor implements PyProjectProcessor {
...
+      const execOptions: ExecOptions = {
+        docker: {},
+        toolConstraints: [toolConstraint],
+      };
+
+      // on lockFileMaintenance do not specify any packages and update the complete lock file
+      // else only update specific packages
+      let packageList = '';
+      if (!isLockFileMaintenance) {
+        packageList = updatedDeps
```
Comment (viceice) [inline lib/modules/manager/pep621/processors/pdm.ts:82]:
```suggestion
        packageList = ' ' + updatedDeps
```

File: lib/modules/manager/pep621/processors/pdm.ts [hunk 7]
```diff
@@ -39,4 +48,75 @@ export class PdmProcessor implements PyProjectProcessor {
...
+          .map((value) => value.packageName ?? value.depName)
+          .join(' ');
+      }
+      const cmd = `pdm update ${packageList}`;
```
Comment (viceice) [inline lib/modules/manager/pep621/processors/pdm.ts:86]:
```suggestion
      const cmd = `pdm update${packageList}`;
```

no trailing spaces

File: lib/modules/manager/pep621/processors/pdm.ts [hunk 8]
```diff
@@ -39,4 +48,75 @@ export class PdmProcessor implements PyProjectProcessor {
...
+            type: 'addition',
+            path: lockFileName,
+            contents: newLockContent,
+          },
+        });
+      } else {
+        logger.debug('pdm.lock is unchanged');
+      }
+
+      return fileChanges.length > 0 ? fileChanges : null;
```
Comment (viceice) [inline lib/modules/manager/pep621/processors/pdm.ts:105]:
```suggestion
      return fileChanges.length ? fileChanges : null;
```

simplify

File: lib/modules/manager/pep621/processors/pdm.ts [hunk 9]
```diff
@@ -39,4 +48,75 @@ export class PdmProcessor implements PyProjectProcessor {
...
+      // on lockFileMaintenance do not specify any packages and update the complete lock file
+      // else only update specific packages
+      let packageList = '';
+      if (!isLockFileMaintenance) {
+        packageList = updatedDeps
+          .map((value) => value.packageName ?? value.depName)
+          .join(' ');
+      }
+      const cmd = `pdm update ${packageList}`;
+      await exec([cmd], execOptions);
```
Comment (viceice) [inline lib/modules/manager/pep621/processors/pdm.ts:87]:
```suggestion
      await exec(cmd, execOptions);
```
you can pass a string as single command

**Review by viceice — CHANGES_REQUESTED** [review]

**Review by viceice — DISMISSED** [review]

File: lib/modules/manager/pep621/artifacts.ts (hunk 2 above)
Comment (secustor) [reply lib/modules/manager/pep621/artifacts.ts:11]: Composer and Bundler for example do this.

**Review by rarkins — APPROVED** [review]

File: lib/modules/manager/pep621/artifacts.ts (hunk 2 above)
Comment (viceice) [reply lib/modules/manager/pep621/artifacts.ts:11]: we should remove those all. as said, renovate writes all updated package files before calling updateArtifacts.

so this is slowing down renovate unnecessarily.

**Pull request merged**